lib/rules/redux-no-bound-selectors.js

```diff
@@ -0,0 +1,166 @@
+		 * tell which state-aware functions are actually forbidden from
+		 * including binding/instantiation of new functions, we progressively
+		 * store them in this collection.
+		 *
+		 */
+		const reports = [];
+
+		/*
+		 * A collection of bodies of [Arrow]FunctionExpressions that ever
+		 * become the first argument of `connect`. In a typical module, this
+		 * array will contain 0 or 1 elements.
+		 *
+		 * In Program:exit, any element in `reports` whose `outer` property is
+		 * contained in `nodesToReportOn` will yield an actual eslint report.
+		 */
+		const nodesToReportOn = [];
+
+		/*
+		 * Naming is hard. ¯\_(ツ)_/¯
+		 *
+		 * Refers to a body of a function that introduces `state` (via its
+		 * arguments) into the subsequent closures.
+		 *
+		 * When traversing the AST, if this value is falsy we can assume that
+		 * we are not in a closure whose broader scope contains `state`.
+		 */
+		let stateProviderFunctionBody;
+
+		function onFunctionEnter( node ) {
+			if ( stateProviderFunctionBody ) {
+				reports.push( {
+					inner: node,
+					outer: stateProviderFunctionBody,
+					message: FUNC_ERROR_MESSAGE,
+				} );
+				return;
+			}
+
+			if ( isStateInArgs( node ) ) {
+				stateProviderFunctionBody = node.body;
+			}
+		}
+
+		function onFunctionExit( node ) {
+			if ( stateProviderFunctionBody && isStateInArgs( node ) ) {
+				stateProviderFunctionBody = null;
+			}
+		}
+
+		function onIdentifierEnter( node ) {
+			if ( stateProviderFunctionBody && node.name === 'bind' ) {
+				reports.push( {
+					inner: node,
+					outer: stateProviderFunctionBody,
+					message: BIND_ERROR_MESSAGE,
+				} );
+			}
+		}
+
+		return {
+			'Program:exit': function() {
+				reports
+					.filter( ( { outer } ) =>
+						contains( nodesToReportOn, outer ) )
+					.forEach( ( { inner, message } ) =>
+						context.report( inner, message ) );
+			},
+
+			CallExpression: function( node ) {
+				if ( node.callee.name !== 'connect' ) {
+					return;
+				}
+
+				/*
+				 * The first argument of `connect` is expected to be either a
+				 * variable for a function, or an anonymous/inlined function.
+				 * Let's handle both cases.
+				 */
+				const mapStateNode = node.arguments[ 0 ];
+
+				let mapStateBody;
+
+				if ( mapStateNode.type === 'Identifier' ) {
+					/*
+					 * If it's a variable, obtain the function it refers to,
+					 * which should be readily available in scope.
+					 */
+					const mapStateName = mapStateNode.name;
+					const variable = context.getScope().set.get( mapStateName );
+					const definition = variable && variable.defs && variable.defs[ 0 ];
+					mapStateBody = definition && definition.node &&
+						definition.node.init &&
+						definition.node.init.body;
+				} else {
+					/*
+					 * Inlined function? Easy peasy.
+					 */
+					mapStateBody = mapStateNode.body;
+				}
+
+				if ( mapStateBody ) {
+					nodesToReportOn.push( mapStateBody );
+				}
+			},
+
+			Identifier: onIdentifierEnter,
```

Comment:
I'm a little worried about the performance implications of an Identifier visitor. Do you think `CallExpression.callee.property.name` might be reasonable to check (and potentially more accurate) instead?

![image](https://cloud.githubusercontent.com/assets/1779930/26510873/42008f76-422d-11e7-8b56-3f093726413c.png)

http://astexplorer.net/#/gist/1ce6f31efbd822108c1227fbb336098a/7c73e810541bec4f0f94dd486de21f49198ca697